### ISSUE TITLE
CA1508: Dead code false positive

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn-analyzers#3645]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

Example: [CA1508](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1508)

#### Repro steps

We are using some third-party library which doesn't use nullable code but we do use it on our codebase. The code is the following:

```
#pragma warning disable CA1508 // Avoid dead conditional code - False positive
var annotationKind = annotation.Tag as AnnotationKind?
#pragma warning restore CA1508 // Avoid dead conditional code
    ?? throw new InvalidOperationException("Unexpected tag in pasted annotations.");
```

**Comment by Evangelink** [comment]
I have added a unit-test and I can confirm that even without the nullable context there is a FP for the rule.

```csharp
public enum SomeEnum { A, }

public class MyClass
{
    public object Tag { get; set; }

    public void M(MyClass c)
    {
        var result = c.Tag as SomeEnum?
            ?? throw new System.InvalidOperationException(""some message"");
    }
}
```

**Comment by lioobayoyo** [comment]
Similar issue here, here a reproducible example. Using the latest nightly build of FxCopAnalyzer

We're on .NET 4.8, but the issue persists if we use  `<TargetFramework>netcoreapp3.1</TargetFramework>`

Code
==

.csproj
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net48</TargetFramework>
    <LangVersion>8.0</LangVersion>
    <Nullable>enable</Nullable>
    <Platforms>x64;x86</Platforms>
  </PropertyGroup>

  <ItemGroup>
    <None Include="..\.editorconfig" Link=".editorconfig" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.3.0-beta3.20404.9">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="System.Collections.Immutable" Version="1.7.1" />
  </ItemGroup>
</Project>
```

code
```
using System.Collections.Generic;
using System.Collections.Immutable;

namespace ConsoleApp1
{
    public class Foo
    {
        public Foo(IEnumerable<object> enumerable)
        {
            // Warning on the `as` expression.
            ImmutableArray = enumerable as ImmutableArray<object>?
                ?? enumerable.ToImmutableArray();
        }

        public ImmutableArray<object> ImmutableArray { get; }
    }
}
```

.editorconfig
```
[*.cs]

# CA1508: Avoid dead conditional code
dotnet_diagnostic.CA1508.severity = warning
```
error:
```
Warning	CA1508	'enumerable as ImmutableArray<object>?' is never 'null'. Remove or refactor the condition(s) to avoid dead code.	ConsoleApp1	.../ConsoleApp1\Foo.cs	10	Active
```

Note:
==

There is a workaround by using `is` : 

```
// No false warning here.
ImmutableArray = enumerable is ImmutableArray<object> alreadyGood ? alreadyGood : enumerable.ToImmutableArray();
```

**Comment by Evangelink** [comment]
@mavasani It'd be nice if we could make some progress on the various `CA1508` related tickets.

**Comment by langdonx** [comment]
> @mavasani It'd be nice if we could make some progress on the various `CA1508` related tickets.

17 of 'em now. ;)

**Comment by mavasani** [comment]
Verified none of the repros mentioned in this issue repro anymore, closing out.
